### PR TITLE
deploy hook with salt for deterministic address

### PR DESCRIPTION
**Pull request by mejango**
# Description

deploy hook with salt for deterministic address

## Limitations & risks

*Are there any trade-off or new vulnarbility surface based on theses changes?*

# Check-list
- [ ] Tests are covering the new feature
- [ ] Code is [natspec'd](https://docs.soliditylang.org/en/v0.8.17/natspec-format.html)
- [ ] Code is [linted and formatted](https://docs.soliditylang.org/en/v0.8.17/style-guide.html)
- [ ] I have run the test locally (and they pass)
- [ ] I have rebased to the latest main commit (and tests still pass)

# Interactions
These changes will impact the following contracts:
- Directly:

- Indirectly:

### CODE REVIEW

File: src/JB721TiersHookDeployer.sol [hunk 1]
```diff
@@ -61,17 +61,21 @@ contract JB721TiersHookDeployer is ERC2771Context, IJB721TiersHookDeployer {
...
     )
         external
         override
         returns (IJB721TiersHook newHook)
     {
         // Deploy the governance variant specified by the config.
-        newHook = IJB721TiersHook(Clones.clone(address(HOOK)));
+        newHook = IJB721TiersHook(
+            salt == bytes32(0) ? Clones.clone(address(HOOK)) : Clones.cloneDeterministic(address(HOOK), salt)
+        );
```
Comment (xBA5ED) [inline src/JB721TiersHookDeployer.sol:84]: If the salt is set we should hash it combined with the `msg.sender`  to make sure only the sender can deploy to a certain address. ([for reference](https://github.com/Bananapus/nana-core/blob/b5dbcdd0a3bad6c8b5830718ee8dcd5094606b84/src/JBTokens.sol#L253-L255)]

File: src/JB721TiersHookProjectDeployer.sol [hunk 2]
```diff
@@ -78,7 +80,7 @@ contract JB721TiersHookProjectDeployer is JBPermissioned, IJB721TiersHookProject
         projectId = DIRECTORY.PROJECTS().count() + 1;
 
         // Deploy the hook.
-        hook = HOOK_DEPLOYER.deployHookFor(projectId, deployTiersHookConfig);
+        hook = HOOK_DEPLOYER.deployHookFor(projectId, deployTiersHookConfig, salt);
```
Comment (xBA5ED) [inline src/JB721TiersHookProjectDeployer.sol]: We should also hash it here with the `msg.sender`, this way we make a path using hashes that needs to be followed to get the same deployment address for `721` ([for reference](https://github.com/Bananapus/nana-core/blob/b5dbcdd0a3bad6c8b5830718ee8dcd5094606b84/src/JBController.sol#L462)). Same thing for `launchRulesetsFor` and `queueRulesetsOf`

File: src/JB721TiersHookDeployer.sol (hunk 1 above)
Comment (mejango) [reply src/JB721TiersHookDeployer.sol:84]: Fixed

File: src/JB721TiersHookProjectDeployer.sol (hunk 2 above)
Comment (mejango) [reply src/JB721TiersHookProjectDeployer.sol]: Fixed

File: src/JB721TiersHookProjectDeployer.sol [hunk 3]
```diff
@@ -114,7 +118,7 @@ contract JB721TiersHookProjectDeployer is JBPermissioned, IJB721TiersHookProject
         });
 
         // Deploy the hook.
-        hook = HOOK_DEPLOYER.deployHookFor(projectId, deployTiersHookConfig);
+        hook = HOOK_DEPLOYER.deployHookFor(projectId, deployTiersHookConfig, salt);
```
Comment (xBA5ED) [inline src/JB721TiersHookProjectDeployer.sol]: Another one here

File: src/JB721TiersHookProjectDeployer.sol (hunk 2 above)
Comment (xBA5ED) [reply src/JB721TiersHookProjectDeployer.sol]: the line/function didn't change :eyes:

**Pull request merged**